source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -0,0 +1,214 @@
+  }
+
+  if (status == TSI_OK && handshaker_result != nullptr) {
+    tsi_peer peer;
+    tsi_handshaker_result_extract_peer(handshaker_result, &peer);
+    ENVOY_CONN_LOG(debug, "TSI: Handshake successful: peer properties: {}",
+                   callbacks_->connection(), peer.property_count);
+    for (size_t i = 0; i < peer.property_count; ++i) {
+      ENVOY_CONN_LOG(debug, "  {}: {}", callbacks_->connection(), peer.properties[i].name,
+                     std::string(peer.properties[i].value.data, peer.properties[i].value.length));
+    }
+    if (handshake_validator_) {
+      std::string err;
+      bool peer_validated = handshake_validator_(peer, err);
+      if (peer_validated) {
+        ENVOY_CONN_LOG(info, "TSI: Handshake validation succeeded.", callbacks_->connection());
+      } else {
+        ENVOY_CONN_LOG(warn, "TSI: Handshake validation failed: {}", callbacks_->connection(), err);
+        tsi_peer_destruct(&peer);
+        return Network::PostIoAction::Close;
+      }
+    } else {
+      ENVOY_CONN_LOG(info, "TSI: Handshake validation skipped.", callbacks_->connection());
+    }
+    tsi_peer_destruct(&peer);
+
+    const unsigned char* unused_bytes;
+    size_t unused_byte_size;
+
+    status =
+        tsi_handshaker_result_get_unused_bytes(handshaker_result, &unused_bytes, &unused_byte_size);
+    ASSERT(status == TSI_OK);
+    if (unused_byte_size > 0) {
+      raw_read_buffer_.add(unused_bytes, unused_byte_size);
+    }
+    ENVOY_CONN_LOG(debug, "TSI: Handshake successful: unused_bytes: {}", callbacks_->connection(),
+                   unused_byte_size);
+
+    tsi_frame_protector* frame_protector;
+    status =
+        tsi_handshaker_result_create_frame_protector(handshaker_result, NULL, &frame_protector);
+    ASSERT(status == TSI_OK);
+    frame_protector_ = std::make_unique<TsiFrameProtector>(frame_protector);
+
+    handshake_complete_ = true;
+    callbacks_->raiseEvent(Network::ConnectionEvent::Connected);
+  }
+
+  if (raw_read_buffer_.length() > 0) {
+    callbacks_->setReadBufferReady();
+  }
+
+  // Try to write raw buffer when next call is done, even this is not in do[Read|Write] stack.
+  if (raw_write_buffer_.length() > 0) {
+    return raw_buffer_socket_->doWrite(raw_write_buffer_, false).action_;
+  }
+
+  return Network::PostIoAction::KeepOpen;
+}
+
+Network::IoResult TsiSocket::doRead(Buffer::Instance& buffer) {
+  Network::IoResult result = raw_buffer_socket_->doRead(raw_read_buffer_);
+  ENVOY_CONN_LOG(debug, "TSI: raw read result action {} bytes {} end_stream {}",
+                 callbacks_->connection(), enumToInt(result.action_), result.bytes_processed_,
+                 result.end_stream_read_);
+  if (result.action_ == Network::PostIoAction::Close && result.bytes_processed_ == 0) {
+    return result;
+  }
+
+  if (!handshake_complete_) {
+    Network::PostIoAction action = doHandshake();
+    if (action == Network::PostIoAction::Close || !handshake_complete_) {
```

Comment:
Is this condition ever satisfied? It seems that doHandshake() always return Network::PostIoAction::KeepOpen.

Reply:
I'll leave this uncovered until next PR to follow up on synchronous TSI optimization.

Reply:
Ah, that's for synchronous handshake! Makes sense.